mono/metadata/w32socket-unix.c

```diff
@@ -943,12 +943,18 @@ mono_w32socket_setsockopt (SOCKET sock, gint level, gint optname, gconstpointer
 #if defined (__linux__)
 	else if (level == SOL_SOCKET &&
 		   (optname == SO_SNDBUF || optname == SO_RCVBUF)) {
+		
+		/*
+	         * bufsize = *((int *) optval); is undefined behaviour, so we use memcpy.
```

Comment:
It's defined behavior.

mono_w32socket_setsockopt for SO_RCVBUF is called from here:

https://github.com/mono/mono/blob/9a5a55505d85d332c0233a111539d5ea9834f8ce/mono/metadata/w32socket.c#L2187

so the address passed in `optval` is an int.  casting a pointer to void* and back to the same type is absolutely defined behavior.

optval is a pointer to

Reply:
gint
mono_w32socket_setsockopt (SOCKET sock, gint level, gint optname, gconstpointer optval, socklen_t optlen)
{
'''
       int bufsize = *((int *) optval);
}
I think, gcc doesn't care about where "optval" come from, only care about "optval" parameter declarations.
because "mono_w32socket_setsockopt" may call from anywhere, in gcc compilation context
I also find Strict Aliasing Rule in google, 
https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8
https://www.approxion.com/pointers-c-part-iii-strict-aliasing-rule/
may be help.

Reply:
Thanks for the references.   In this case, the *effective type* of `int_val` is `int`.  It is passed as an argument of type `void*` but it is accessed again as an lvalue with type `int`.  So this is defined behavior.

Effectively the code is:

```c
int int_val;
const void* optval = &int_val;
int bufsize = *((int*)optval);  /* lvalue has type int, so this is not UB */
```

the lvalue `*((int*)p)` is used with the type `int` - the same as the effective type of the storage.  https://en.cppreference.com/w/c/language/object#Effective_type

Reply:
Sorry, i held up by something. 

in url 
https://en.cppreference.com/w/c/language/object#Effective_type

Given an object with effective type T1, using an lvalue expression (typically, dereferencing a pointer) of a different type T2 is undefined behavior, unless:

T2 and T1 are compatible types.  (int* and void * not compatible types)
T2 is cvr-qualified version of a type that is compatible with T1. (no )
T2 is a signed or unsigned version of a type that is compatible with T1. (no)
T2 is an aggregate type or union type type that includes one of the aforementioned types among its members (including, recursively, a member of a subaggregate or contained union). (no unicon)
T2 is a character type (char, signed char, or unsigned char). (no char)

so, i think it`s still undefined behavior.

Effectively code 

int int_val;
const void* optval = &int_val;
int bufsize = *((int*)optval);  /* lvalue has type int, so this is not UB */

try it in gcc  O2,  

in issues #20765

no problem in gcc -g
only has problem in gcc -O2
if build with gcc -O2 -fno-strict-aliasing it`s good.

Reply:
Ok, tried it with various versions of GCC and Clang https://godbolt.org/z/j8d5Tf

In all cases the round trip `*(int*)(void*)&int` is well-defined and compiles away entirely.  There is no pointer aliasing going on here.

I'm sorry but I don't think this is the right fix for the issue that you are seeing.

Reply:
Maybe I'm wrong..I'm sorry for your time.
emm,c is a complicated language,I need to learn more.
Thanks a lot.